File: BlockOfLabs2/laboratory2.2/lab2.2.3.cpp
#include <iostream>
using namespace std;


const int MAX_N = 1000000;
const int MAX_K = 100000;
const int MIN_K = 1;


int input() {
	int k;
	bool isIncorrect = true;
	do
	{
		cin >> k;
		if (cin.fail() || cin.get() != '\n') {
			cout << "Invalid numeric input.\n";
			cin.clear();
			while (cin.get() != '\n');
		}
		else if (k < MIN_K || k > MAX_K)
			cout << "Number should be from " << MIN_K << " to " << MAX_K << ".\n";
		else
			isIncorrect = false;
	} while (isIncorrect);

	return k;
}


int sumOfDigits(int num) {
	int sum = 0;
	while (num) {
		sum += num % 10;
		num /= 10;
	}
	return sum;
}


bool checkSum(int Sum, int k, int nutNumb) {
	return k * Sum == nutNumb;
}


void searchNum(int k) {
	int nutNumb = k;
	while (nutNumb <= MAX_N) {
		int sum = sumOfDigits(nutNumb);
		if (checkSum(sum, k, nutNumb))
			cout << nutNumb << " ";
		nutNumb += k;
	}
}


int main()
{
	cout << "The program finds all natural numbers that are k times the sum of their digits.\n";

	cout << "Write K number from " << MIN_K << " to " << MAX_K << ":\n";
	int k = input();

	searchNum(k);
	return 0;
}

File: BlockOfLabs2/laboratory2.2/lab2.2.3.java
package lab2;

import java.util.Scanner;

public class lab2 {
    static final int MAX_N = 1000000;
    static final int MAX_K = 100000;
    static final int MIN_K = 1;


    static int input() {
        Scanner in = new Scanner(System.in);
        int k = 0;
        boolean isIncorrect = true;
        do {
            try {
                k = Integer.parseInt(in.nextLine());
            } catch (NumberFormatException error) {
                System.err.print("Invalid numeric input.\n");
            }
            if (k < MIN_K || k > MAX_K)
                System.err.printf("Number should be from %d to %d.\n", MIN_K, MAX_K);
            else
                isIncorrect = false;
        } while (isIncorrect);
        in.close();

        return k;
    }


    static int sumOfDigits(int num) {
        int sum = 0;
        while (num >= 1) {
            sum += num % 10;
            num /= 10;
        }
        return sum;
    }


    static boolean checkSum(int Sum, int k, int nutNumb) {
        return k * Sum == nutNumb;
    }


    static void searchNum(int k) {
        int nutNumb = k;
        while (nutNumb <= MAX_N){
            int Sum = sumOfDigits(nutNumb);
            if(checkSum(Sum, k, nutNumb)) {
                System.out.printf("%d ", nutNumb);
            }
            nutNumb += k;
        }
    }


    public static void main(String[] args) {
        System.out.print("The program finds all natural numbers that are k times the sum of their digits.\n");

        System.out.printf("Write K number from %d to %d:\n", MIN_K, MAX_K);
        int k = input();

        searchNum(k);
    }
}